pkg/common/cns-lib/volume/manager.go

```diff
@@ -3402,6 +3406,202 @@ func (m *defaultManager) getAggregatedSnapshotSize(ctx context.Context, volumeID
+	return batchAttachResult, nil
+}
+
+// constructBatchAttachSpecList constructs the CnsVolumeAttachDetachSpec list by
+// going through all volumes batchAttachRequest.
+func constructBatchAttachSpecList(ctx context.Context, vm *cnsvsphere.VirtualMachine,
+	batchAttachRequest []BatchAttachRequest) ([]cnstypes.CnsVolumeAttachDetachSpec, error) {
+	log := logger.GetLogger(ctx)
+	var cnsAttachSpecList []cnstypes.CnsVolumeAttachDetachSpec
+	for _, volume := range batchAttachRequest {
+		// Initialise cnsAttachDetachSpec
+		cnsAttachDetachSpec := cnstypes.CnsVolumeAttachDetachSpec{
+			VolumeId: cnstypes.CnsVolumeId{
+				Id: volume.VolumeID,
+			},
+			Vm:       vm.Reference(),
+			Sharing:  volume.SharingMode,
+			DiskMode: volume.DiskMode,
+		}
+
+		// Set controllerKey and unitNumber only if they are provided by the user.
+		if volume.ControllerKey != "" {
+			// Convert to int64
+			controllerKey, err := strconv.ParseInt(volume.ControllerKey, 10, 64)
+			if err != nil {
+				log.Errorf("failed to convert controllerKey %s to integer", controllerKey)
+				return cnsAttachSpecList, err
+			}
+			cnsAttachDetachSpec.ControllerKey = controllerKey
+		}
+		if volume.UnitNumber != "" {
+			// Convert to int64
+			unitNumber, err := strconv.ParseInt(volume.UnitNumber, 10, 64)
+			if err != nil {
+				log.Errorf("failed to convert unitNumber %s to integer", unitNumber)
+				return cnsAttachSpecList, err
+			}
+			cnsAttachDetachSpec.UnitNumber = unitNumber
+		}
+		cnsAttachSpecList = append(cnsAttachSpecList, cnsAttachDetachSpec)
+	}
+
+	log.Infof("Constucted CnsVolumeAttachDetachSpec for VM %s", vm.UUID)
+	return cnsAttachSpecList, nil
+
+}
+
+// BatchAttachVolumes calls CNS Attach Volume API with a list of volumes for a single VM.
+func (m *defaultManager) BatchAttachVolumes(ctx context.Context,
+	vm *cnsvsphere.VirtualMachine,
+	batchAttachRequest []BatchAttachRequest) ([]BatchAttachResult, string, error) {
+	ctx, cancelFunc := ensureOperationContextHasATimeout(ctx)
+	defer cancelFunc()
+	internalBatchAttachVolumes := func() ([]BatchAttachResult, string, error) {
+		log := logger.GetLogger(ctx)
+		var faultType string
+		batchAttachResult := make([]BatchAttachResult, 0)
+
+		log.Infof("Starting batch attach for VM %s", vm.UUID)
+
+		err := validateManager(ctx, m)
+		if err != nil {
+			log.Errorf("failed to validate manger. Err: %s", err)
+			faultType = ExtractFaultTypeFromErr(ctx, err)
+			return batchAttachResult, faultType, err
+		}
+		// Set up the VC connection.
+		err = m.virtualCenter.ConnectCns(ctx)
+		if err != nil {
+			log.Errorf("ConnectCns failed with err: %s", err)
+			faultType = ExtractFaultTypeFromErr(ctx, err)
+			return batchAttachResult, faultType, err
+		}
+
+		// Construct the CNS AttachSpec list.
+		cnsAttachSpecList, err := constructBatchAttachSpecList(ctx, vm, batchAttachRequest)
+		if err != nil {
+			log.Errorf("CNS AttachVolume failed from vCenter %q with err: %v", m.virtualCenter.Config.Host, err)
```

Comment:
Can you change error message to say - "constructBatchAttachSpecList failed with err"